File: Day100/Part6/self_healing_string.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

// Problem 6: Self-Healing String
// "Can the substring S[l...r] be rearranged to form a palindrome?"
// AND: "Update S[i] = c".
//
// Logic: A string can be rearranged into a palindrome iff
// at most one character appears an odd number of times.
// We maintain the parity of character counts for each substring using a Segment Tree.
//
// State: A bitmask of length 26.
// 'a' flips bit 0, 'b' flips bit 1, ...
// XOR sum processing. 
// Query(l, r) -> get XOR sum of range.
// Check popcount(mask) <= 1.
//
// Complexity: O(log N) per update/query. 26 bits fits in int.

struct SegmentTree {
    int n;
    string S;
    vector<int> tree;

    SegmentTree(string s) : S(s) {
        n = s.length();
        tree.resize(4 * n);
        build(1, 0, n - 1);
    }

    void build(int node, int start, int end) {
        if (start == end) {
            tree[node] = (1 << (S[start] - 'a'));
        } else {
            int mid = (start + end) / 2;
            build(2 * node, start, mid);
            build(2 * node + 1, mid + 1, end);
            tree[node] = tree[2 * node] ^ tree[2 * node + 1];
        }
    }

    void update(int idx, char new_val) {
        updatePtr(1, 0, n - 1, idx, new_val);
    }

    void updatePtr(int node, int start, int end, int idx, char new_val) {
        if (start == end) {
            S[idx] = new_val;
            tree[node] = (1 << (S[idx] - 'a'));
        } else {
            int mid = (start + end) / 2;
            if (start <= idx && idx <= mid) {
                updatePtr(2 * node, start, mid, idx, new_val);
            } else {
                updatePtr(2 * node + 1, mid + 1, end, idx, new_val);
            }
            tree[node] = tree[2 * node] ^ tree[2 * node + 1];
        }
    }

    int query(int l, int r) {
        return queryPtr(1, 0, n - 1, l, r);
    }

    int queryPtr(int node, int start, int end, int l, int r) {
        if (r < start || end < l) return 0;
        if (l <= start && end <= r) return tree[node];
        int mid = (start + end) / 2;
        int p1 = queryPtr(2 * node, start, mid, l, r);
        int p2 = queryPtr(2 * node + 1, mid + 1, end, l, r);
        return p1 ^ p2;
    }
};

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int N, Q;
    if (!(cin >> N >> Q)) return 0;
    
    string S;
    cin >> S;

    SegmentTree st(S);

    for(int i=0; i<Q; ++i) {
        int type;
        cin >> type;
        if (type == 1) {
            int idx;
            char c;
            cin >> idx >> c;
            // 1-based index input usually? Let's assume 1-based for problem consistency
            st.update(idx - 1, c);
        } else {
            int l, r;
            cin >> l >> r;
            int mask = st.query(l - 1, r - 1);
            // Check popcount
            // Built-in popcount
            int odd_counts = 0;
            while(mask > 0) {
                if (mask & 1) odd_counts++;
                mask >>= 1;
            }
            if (odd_counts <= 1) cout << "YES\n";
            else cout << "NO\n";
        }
    }
    return 0;
}

File: Day100/Part6/self_healing_string.py
import sys

# Python Segment Tree for "Self-Healing String"
# Operations:
# 1 idx char: Update S[idx] = char
# 2 l r: Can S[l...r] be rearranged to palindrome? (Popcount(XOR) <= 1)

class SegmentTree:
    def __init__(self, s):
        self.n = len(s)
        self.s = list(s)
        self.tree = [0] * (4 * self.n)
        self._build(1, 0, self.n - 1)

    def _get_mask(self, char):
        return 1 << (ord(char) - ord('a'))

    def _build(self, node, start, end):
        if start == end:
            self.tree[node] = self._get_mask(self.s[start])
        else:
            mid = (start + end) // 2
            self._build(2 * node, start, mid)
            self._build(2 * node + 1, mid + 1, end)
            self.tree[node] = self.tree[2 * node] ^ self.tree[2 * node + 1]

    def update(self, idx, char):
        self._update(1, 0, self.n - 1, idx, char)

    def _update(self, node, start, end, idx, char):
        if start == end:
            self.s[idx] = char
            self.tree[node] = self._get_mask(char)
        else:
            mid = (start + end) // 2
            if start <= idx <= mid:
                self._update(2 * node, start, mid, idx, char)
            else:
                self._update(2 * node + 1, mid + 1, end, idx, char)
            self.tree[node] = self.tree[2 * node] ^ self.tree[2 * node + 1]

    def query(self, l, r):
        return self._query(1, 0, self.n - 1, l, r)

    def _query(self, node, start, end, l, r):
        if r < start or end < l:
            return 0
        if l <= start and end <= r:
            return self.tree[node]
        mid = (start + end) // 2
        p1 = self._query(2 * node, start, mid, l, r)
        p2 = self._query(2 * node + 1, mid + 1, end, l, r)
        return p1 ^ p2

def solve():
    input_data = sys.stdin.read().split()
    if not input_data: return
    iterator = iter(input_data)
    
    try:
        N = int(next(iterator))
        Q = int(next(iterator))
        S_str = next(iterator)
        
        st = SegmentTree(S_str)
        
        results = []
        for _ in range(Q):
            type = int(next(iterator))
            if type == 1:
                idx = int(next(iterator))
                c = next(iterator)
                st.update(idx - 1, c)
            else:
                l = int(next(iterator))
                r = int(next(iterator))
                mask = st.query(l-1, r-1)
                
                # Popcount
                odd = bin(mask).count('1')
                results.append("YES" if odd <= 1 else "NO")
                
        sys.stdout.write('\n'.join(results) + '\n')
            
    except StopIteration:
        return

if __name__ == "__main__":
    solve()
